core/arch/arm/kernel/thread_spmc.c

```diff
@@ -326,46 +327,111 @@ static bool is_my_uuid(uint32_t w0, uint32_t w1, uint32_t w2, uint32_t w3)
 	       w2 == 0x02005ebc && w3 == 0x1bc5d5a5;
 }
 
-static void handle_partition_info_get(struct thread_smc_args *args,
-				      struct ffa_rxtx *rxtx)
+void spmc_fill_partition_entry(struct ffa_partition_info *fpi,
+			       uint16_t endpoint_id, uint16_t execution_context)
 {
-	uint32_t ret_fid = 0;
-	int rc = 0;
+	fpi->id = endpoint_id;
+	/* Number of execution contexts implemented by this partition */
+	fpi->execution_context = execution_context;
 
-	if (!is_nil_uuid(args->a1, args->a2, args->a3, args->a4) &&
-	    !is_my_uuid(args->a1, args->a2, args->a3, args->a4)) {
-		ret_fid = FFA_ERROR;
-		rc = FFA_INVALID_PARAMETERS;
-		goto out;
+	fpi->partition_properties = FFA_PARTITION_DIRECT_REQ_RECV_SUPPORT |
+				    FFA_PARTITION_DIRECT_REQ_SEND_SUPPORT;
+}
+
+static uint32_t handle_partition_info_get_all(size_t *elem_count,
+					      struct ffa_rxtx *rxtx)
+{
+	struct ffa_partition_info *fpi = rxtx->tx;
+
+	/* Add OP-TEE SP */
+	spmc_fill_partition_entry(fpi, my_endpoint_id, CFG_TEE_CORE_NB_CORE);
+	rxtx->tx_is_mine = false;
+	*elem_count = 1;
+	fpi++;
+
+	if (IS_ENABLED(CFG_SECURE_PARTITION)) {
+		size_t count = (rxtx->size / sizeof(*fpi)) - 1;
+
+		if (sp_partition_info_get_all(fpi, &count))
+			return FFA_NO_MEMORY;
+		*elem_count += count;
 	}
 
-	cpu_spin_lock(&rxtx->spinlock);
-	if (rxtx->size && rxtx->tx_is_mine) {
-		struct ffa_partition_info *fpi = rxtx->tx;
+	return FFA_OK;
+}
 
-		fpi->id = my_endpoint_id;
-		fpi->execution_context = CFG_TEE_CORE_NB_CORE;
-		/*
-		 * Supports receipt of direct requests.
-		 * Can send direct requests.
-		 */
-		fpi->partition_properties = BIT(0) | BIT(1);
+void spmc_handle_partition_info_get(struct thread_smc_args *args,
+				    struct ffa_rxtx *rxtx)
+{
+	uint32_t ret_fid = FFA_ERROR;
+	uint32_t rc = 0;
+	uint32_t endpoint_id = my_endpoint_id;
+	struct ffa_partition_info *fpi = NULL;
+	uint16_t execution_context = CFG_TEE_CORE_NB_CORE;
 
-		ret_fid = FFA_SUCCESS_32;
-		rc = 1;
-		rxtx->tx_is_mine = false;
-	} else {
-		ret_fid = FFA_ERROR;
+	cpu_spin_lock(&rxtx->spinlock);
+
+	if (!rxtx->size || !rxtx->tx_is_mine) {
 		if (rxtx->size)
 			rc = FFA_BUSY;
 		else
 			rc = FFA_DENIED; /* TX buffer not setup yet */
+		goto out;
+	}
+
+	fpi = rxtx->tx;
```

Comment:
`handle_partition_info_get_all()` should check at least `rxtx->size >= sizeof(*fpi)` since at least 1 cell is filled.
I think caller of `spmc_fill_partition_entry()` at lines 401 and 420 should also check `rxtx` is sized for a 1 cell.

So maybe worth here to check `rxtx->size >= sizeof(*fpi)`.

Reply:
Done